R/indexing.R

```diff
@@ -391,37 +389,31 @@ length.igraph <- function(x) {
     (is.null(attr) && is.numeric(value) && value == 0)) {
     ## Delete edges
     if (missing(i) && missing(j)) {
-      todel <- unlist(x[[, , ..., edges = TRUE]])
+      todel <- seq_len(ecount(x))
     } else if (missing(j)) {
-      todel <- unlist(x[[i, , ..., edges = TRUE]])
+      todel <- unlist(incident(x, v = i, mode = "out"))
     } else if (missing(i)) {
-      todel <- unlist(x[[, j, ..., edges = TRUE]])
+      todel <- unlist(incident(x, v = j, mode = "in"))
     } else {
-      todel <- unlist(x[[i, j, ..., edges = TRUE]])
+      edge_pairs <- expand.grid(i, j)
+      edge_ids <- get_edge_ids(x, c(rbind(edge_pairs[, 1], edge_pairs[, 2])))
```

Comment:
Is this covered by tests?

```suggestion
      edge_ids <- get_edge_ids(x, as.vector(t(edge_pairs)))
```

The interface of `get_edge_ids()` is interesting. Should we extend that to accept two-column data frames?

Reply:
Going through the existing tests, I realize there are some gaps. I will add a set of tests

Reply:
> Is this covered by tests?
> 
Going through the existing tests, I realize there are some gaps. I will add a set of tests for this functionality

Reply:
> The interface of `get_edge_ids()` is interesting. Should we extend that to accept two-column data frames?

It has been bothering me mildly for years as a user that edges need to be supplied as a vector (same with `add_edges()` and `delete_edges()` and probably more). However that's required by the c core. It might be too much of a fundamental change at this point?

Reply:
The `c(rbind(...))` pattern is probably fine, I forgot the semantics for vectors:

``` r
c(rbind(1:3, 4:6))
#> [1] 1 4 2 5 3 6
c(t(data.frame(1:3, 4:6)))
#> [1] 1 4 2 5 3 6
as.vector(t(data.frame(1:3, 4:6)))
#> [1] 1 4 2 5 3 6
```

<sup>Created on 2025-01-19 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

There are two layers here: the C core and the R interface. We should provide an idiomatic R user interface that translates to what the C core needs.

Reply:
`c(rbind(...))` is faster for data frames by an order of magnitude, but not for matrices:

``` r
df <- as.data.frame(cbind(1:30, 4:33))

bench::mark(
  c(t(df)),
  c(rbind(df[, 1], df[, 2])),
  c(rbind(df[[1]], df[[2]]))
)
#> # A tibble: 3 × 6
#>   expression                      min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr>                 <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 c(t(df))                    12.96µs  14.51µs    57926.    74.4KB     34.8
#> 2 c(rbind(df[, 1], df[, 2]))   5.33µs   5.99µs   157537.      576B     47.3
#> 3 c(rbind(df[[1]], df[[2]]))   3.65µs   4.26µs   219083.      576B     43.8

m <- cbind(1:30, 4:33)

bench::mark(
  c(t(m)),
  c(rbind(m[, 1], m[, 2]))
)
#> # A tibble: 2 × 6
#>   expression                    min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr>               <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 c(t(m))                     820ns 983.94ns   964028.      576B     96.4
#> 2 c(rbind(m[, 1], m[, 2]))    984ns   1.15µs   791706.      576B      0
```

<sup>Created on 2025-01-20 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

Draft PR for new UI in https://github.com/igraph/rigraph/pull/1663.